keps/sig-scheduling/3838-pod-mutable-scheduling-directives/README.md

```diff
@@ -0,0 +1,962 @@
+
+At the time of creating the workload, it may not be known on which subset of the nodes (e.g., a zone or 
+a VM type) a workload should run. To control where a workload should run, the controller updates 
+the node selector/affinity of the workload's pods; by doing so, the queue controller is able to influence the 
+scheduler and autoscaler decisions.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+- New calls from a queue controller to the API server to update node affinity/selector. 
+  The mitigation is for such controllers to make a single API call for both updating
+  affinity/selector and removing the scheduling gate.
+
+- App controllers may find it surprising that the pod they created is different from their
+  template. Typically apps don't continusly reconile existing pods
+  to match the template, they react only to change to the pod template in the app's
+  spec. Moreover, the side affects of this proposal is no different from a mutating 
+  webhook that injects affinity onto the pod.
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+The pod update validation logic in the API server already allows updating labels and 
+annotations, it also allows adding new tolerations. What we need to do is to relax the update validation
+on node affinity/selector.
+
+There are two main problems we need to address:
+
+1) Race conditions between updating node affinity/selector constraints and the kube-scheduler when scheduling
+the pod. As mentioned before, the solution is to restrict the updates to pods with scheduling gates. This
+prevents the race condition.
+
+2) Invalidating a decision made by a policy admission webhook. The concern here is that some of those policy
+webhooks validate node affinity/selector on CREATE only and don't trigger that validation on UPDATE because 
+they assumed the fields are immutable. This can be addressed by allowing updates
+to node selector/affinity that only further constrain the selection. 
+
+For `spec.nodeSelector`, only additions are allowed. For `spec.nodeAffinity`, 
+`.requiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms`
+are ORed, while `nodeSelectorTerms[].matchExpressions` list and 
+`nodeSelectorTerms[].fieldExpressions` are ANDed, and so for pods with existing 
+`nodeAffinity` terms, only additions of `NodeSelectorRequirements` to `matchExpressions` 
```

Comment:
this "only append to matchExpressions or fieldExpressions of existing terms" makes sense for existing pods which already have `spec.affinity.nodeAffinity.requiredDuringSchedulingIgnoredDuringExecution` term(s)

since `preferredDuringSchedulingIgnoredDuringExecution` can already be ignored if needed by the scheduler, and it doesn't look like there's a way to tack on additional qualifiers to existing terms in that list, appending to the  preferredDuringSchedulingIgnoredDuringExecution list seems more reasonable